File: Queue/Double-Ended Queue/Array-Based/DEQueueArray.py
class DEQueueArray:

    def __init__(self, n):
        self.SIZE = n
        self.elements = [0] * self.SIZE
        self.FRONT = -1
        self.REAR = -1

    #Check whether the Queue is Empty or not
    def isEmpty(self):
        return self.FRONT == -1

    #check whether the Queue is Full or not
    def isFull(self):
        return self.REAR == self.FRONT - 1 or self.FRONT == 0 and self.REAR == self.SIZE - 1

    #Insert At REAR
    def insertREAR(self, x):

        if self.isFull():
            printf("Queue is Full!")
            return

        print("\nBefore Inserting At REAR: FRONT = {} and REAR = {}".format(self.FRONT, self.REAR))

        if self.isEmpty():

            self.FRONT += 1
            self.REAR += 1
            self.elements[self.REAR] = x

        elif self.REAR == self.SIZE - 1:

            self.REAR = 0
            self.elements[self.REAR] = x

        else:

            self.REAR += 1
            self.elements[self.REAR] = x
        
        print("After Inserting At REAR: FRONT = {} and REAR = {}".format(self.FRONT, self.REAR))


    #Insert At FRONT
    def insertFRONT(self, x):

        if self.isFull():
            print("Queue is Full!")
            return

        print("\nBefore Inserting At FRONT: FRONT = {} and REAR = {}".format(self.FRONT, self.REAR))

        if self.isEmpty():

            self.FRONT += 1
            self.REAR += 1

            self.elements[self.FRONT] = x

        elif self.FRONT == 0:

            self.FRONT = self.SIZE - 1
            self.elements[self.FRONT] = x

        else:

            self.FRONT -= 1
            self.elements[self.FRONT] = x
        
        print("After Inserting At FRONT: FRONT = {} and REAR = {}".format(self.FRONT, self.REAR))


    #Deleting From Front
    def deleteFRONT(self):

        if self.isEmpty():
            print("Queue is Empty!")
            return

        print("\nBefore Deleting FROM FRONT: FRONT = {} and REAR = {}".format(self.FRONT, self.REAR))
        self.elements[self.FRONT] = 0

        if self.FRONT == self.REAR:

            self.FRONT = -1
            self.REAR = -1

        elif self.FRONT == self.SIZE - 1:

            self.FRONT = 0

        else:

            self.FRONT += 1
        
        print("After Deleting FROM FRONT: FRONT = {} and REAR = {}".format(self.FRONT, self.REAR))


    #Deleting From REAR
    def deleteREAR(self):

        if self.isEmpty():
            print("Queue is Empty!")
            return

        print("\nBefore Deleting FROM REAR: FRONT = {} and REAR = {}".format(self.FRONT, self.REAR))
        self.elements[self.REAR] = 0

        if self.FRONT == self.REAR:

            self.FRONT = -1
            self.REAR = -1

        elif self.REAR == 0:

            sel.REAR = self.SIZE - 1

        else:

            self.REAR -= 1
        
        print("\nBefore Deleting FROM REAR: FRONT = {} and REAR = {}".format(self.FRONT, self.REAR))

File: Queue/Double-Ended Queue/Array-Based/DEQueueArray.java
public class DEQueueArray{

    private int SIZE;
    private int elements[];
    private int FRONT;
    private int REAR;

    //Inieialization
    DEQueueArray(int num){
        SIZE = num;
        elements = new int[SIZE];
        FRONT = -1;
        REAR = -1;
    }

    //Check whether the Queue is Empty or not
    public boolean isEmpty(){
        return FRONT == -1;
    }

    //Check whether the Queue is Full or not
    public boolean isFull(){
        return REAR == FRONT - 1 || FRONT == 0 && REAR == SIZE - 1;
    }

    //Displaying
    public void display(){

        if(isEmpty()){
            System.out.println("Queue is Empty!");
            return;
        }

        for(int i = 0; i < SIZE; i++)
            System.out.print(elements[i] + " <- ");

        System.out.println();

    }

    //Insert At REAR
    public void insertREAR(int x){

        if(isFull()){
            System.out.println("Queue is Full!");
            return;
        }

        System.out.println("\nBefore Inserting At REAR: FRONT = " + FRONT + " and REAR = " + REAR);

        if(isEmpty()){

            FRONT++;
            elements[++REAR] = x;

        } else if (REAR == SIZE - 1){

            REAR = 0;
            elements[REAR] = x;

        } else {

            elements[++REAR] = x;

        }

        System.out.println("After Inserting At REAR: FRONT = " + FRONT + " and REAR = " + REAR);

    }

    //Insert At Front
    public void insertFRONT(int x){

        if(isFull()){
            System.out.println("Queue is Full!");
            return;
        }

        System.out.println("\nBefore Inserting At FRONT: FRONT = " + FRONT + " and REAR = " + REAR);

        if(isEmpty()){

            FRONT++;
            REAR++;

            elements[FRONT] = x;

        } else if(FRONT == 0){

            FRONT = SIZE - 1;

            elements[FRONT] = x;

        } else {

            elements[--FRONT] = x;

        }

        System.out.println("After Inserting At FRONT: FRONT = " + FRONT + " and REAR = " + REAR);

    }

    //Deletting From FRONT
    public void deleteFRONT(){

        if(isEmpty()){
            System.out.println("Queue is Empty!");
            return;
        }

        System.out.println("Before Deleting From FRONT: FRONT = " + FRONT + " and REAR = " + REAR);

        elements[FRONT] = 0;

        if(FRONT == REAR){

            FRONT = -1;
            REAR = -1;

        } else if(FRONT == SIZE - 1){

            FRONT = 0;

        } else {

            FRONT++;

        }

        System.out.println("After Deleting From FRONT: FRONT = " + FRONT + " and REAR = " + REAR);

    }

    //Deleting FROM REAR
    public void deleteREAR(){

        if(isEmpty()){
            System.out.println("Queue is Empty!");
            return;
        }

        System.out.println("Before Deleting FROM REAR: FRONT = " + FRONT + " and REAR = " + REAR);

        elements[REAR] = 0;

        if(FRONT == REAR){

            FRONT = -1;
            REAR = -1;

        } else if(REAR == 0){

            REAR = SIZE - 1;

        } else {

            REAR--;

        }

        System.out.println("Before Deleting FROM REAR: FRONT = " + FRONT + " and REAR = " + REAR);

    }

}
